### ISSUE TITLE
Inaccessibility to screenreaders

### ISSUE THREAD

**Issue by StrangeGirlMurph** [ether/etherpad-lite#7255]
I am shocked to find out that etherpad isn't accessible to screenreaders at all. How did no one think about that? It's truly horrible. For me it just cycles through the line numbers. No sight of a skip to content link or anything. I mean there is no point in even listing out everything that's bad. See for yourself. 

Yes making it truly accessible to screenreaders with all the edit functionality is hard but allowing people to actually read the content is a must.

Please work on that.

**Comment by SamTV12345** [comment]
Hi @StrangeGirlMurph Could you ellaborate on where Etherpad is not accessible? Would be great to find out so we can fix it.

**Comment by StrangeGirlMurph** [comment]
Hey @SamTV12345, sadly basically everywhere. I don't feel like giving an accessibility in web dev 101 talk and explain the absolute basics. I would hope someone on the team would have at least some knowledge. There is also: https://developer.mozilla.org/en-US/docs/Learn_web_development/Core/Accessibility
I highly recommend you also just try it out yourself! I find tablets are a great way to test screen reader support. 

Basically is etherpad right now just not navigable for screen readers. Some points:
- To get to the main content you have to go over all editing tools in the header. Thats 18 swipes. Thats what skip to content buttons are there for!
- On etherpad.wikimedia.org you can't individually cycle through the paragraphs and links. You have know where they are located in order to click them. It just treats all the text as one big block.
- After that it individually cycles through every single line number and just reads them out loud. Of course a blind person would have no clue that the numbers read are line numbers.

These are just some things. Someone really needs to actually think this through. Your LLM of choice can also offer some directions.

**Comment by StrangeGirlMurph** [comment]
Okay I took the time to make a small screen recording of me just swiping through. People who know what to look for will find many problems from that. Note that doesn't even cover all the problems with actually interacting with the buttons. The chat for example in the end is just not available. 

https://github.com/user-attachments/assets/f94097fd-addd-417f-9888-cd1df0ae604b

**Comment by JohnMcLear** [comment]
We put a lot of work into accessibility with aria..  Is aria not working or something?

**Comment by StrangeGirlMurph** [comment]
Sorry for the delay. Was busy with an exam.
Now: The navigation and using divs and spans for text are the problem! The aria tags seems fine from a quick look.
As you can hear in the video there is no skip to content link so you manually have to cycle through all the tools to get to the text. The text is in spans in divs. The screenreader just sees one block of text. There is no cycling through paragraphs and no way to access links for example. Then there is also this horrible going through all the line numbers. It's all in the video really.

**Comment by JohnMcLear** [comment]
Okay noted, sounds horrific, yea :|

I had to learn about screen readers 10 or so years ago to implement our initial attempt, the web has changed  a lot since so I apologize we didn't update the software to keep up with it!

Do you know how the skip to content might work and be implemented?  If so, please provide docs.

Etherpad doesn't really have the concept of "paragraphs", afaik nor does the web.  For example your comment on Github is just one line of text...  Can you provide a source for information for how screen readers try to resolve this?

I guess with line numbers would you prefer if they are just not part of the screen reading concept?

Happy to make any changes you can suggest if you can provide best practice references :)

**Comment by StrangeGirlMurph** [comment]
Thanks. With paragraphs I mean p tags. They do and should get used. The line numbers should be announced at the start of the line. Until then it would definitely be better for them to just be hidden to screenreaders.
A good start covering the very basics is: 
https://developer.mozilla.org/en-US/docs/Learn_web_development/Core/Accessibility/HTML 
(This is a must read!) (in general are the mdn docs good and extensive)
For skip links in detail: 
https://www.w3schools.com/accessibility/accessibility_skip_links.php

**Comment by JohnMcLear** [comment]
fwiw @StrangeGirlMurph the Mozilla team did the initial accessibility integration so it should be to their specs but I think it's changed quite a bit since the first inception..  

I think this is something that 100% needs doing tho so thanks for your bump in the right direction, we probably took the eye off the ball on accessibility and that's not good enough from a Foundation perspective!

As far as P tags go, I think this might be impossible with our current non-shadow-dom-less approach but I'll certainly look into it :)

**Comment by StrangeGirlMurph** [comment]
A month after those PRs have been merged I just tried it again on my freshly updated instance. The main navigation problem is sadly still there. You still get trapped with the line numbers. There is still no skip to content link. Just wanted to make sure you're still working on this. I wouldn't say that this issue has been completed.

**Comment by JohnMcLear** [comment]
@StrangeGirlMurph When you say freshly updated can you tell me which version or git sha you are on?  If you can please provide exact steps to replicate it might be easier or I can fix in real time and you can test in real time...  I'm in the Discord channel if that's useful..

**Comment by StrangeGirlMurph** [comment]
I am using dockers latest image so I am on the latest 2.7.3 release which includes the relevant commits.

The steps to reproduce are basically identical to the recording I took above: https://github.com/ether/etherpad/issues/7255#issuecomment-3620552642

**Comment by JohnMcLear** [comment]
@StrangeGirlMurph yeah I'm getting closer now, I push to develop and don't do releases(Sam does those), if I bring up a test environment for you accessible online can you test it and feedback?  I'm limited on how much I can test given my environment...

**Comment by JohnMcLear** [comment]
I think I got it, turns out f12 has an accessibility item under elements I didn't know about.....   TIL I guess...

I'll merge to develop and get back on w/ other stuff, we should hopefully release this weekend.

**Comment by StrangeGirlMurph** [comment]
Thanks! I am happy to test if you give me a link

**Comment by StrangeGirlMurph** [comment]
Okay great! There are lots of improvements. Thank you for those!

https://github.com/user-attachments/assets/dacbce25-61c9-4632-bd88-7023ed9aff2b

Other things that I noticed that should be better:
- The biggest problem is that you still can't cycle through the text properly line by line to press links and such. You can only really get to links if you already know where they are and navigate directly to them. There should be videos online on how it should be with regular text.
- The toolbars need a bit of love. Its not very clear to the user when you enter the toolbar and what is all part of it. It would be also nice to have the option to skip the toolbar. They need to be in a semantic element. They seem to be in a ul but something is off. Even lighthouse caught this.
  <img width="793" height="468" alt="Image" src="https://github.com/user-attachments/assets/20fd2978-a298-4e4a-a06a-dce006d541c6" />
  <img width="953" height="686" alt="Image" src="https://github.com/user-attachments/assets/db7772d6-d014-485c-b1aa-7408abe85680" />
- It's not clear what the number next to the users button is
- Between the editor and chat "Ether X" gets announced without context
  <img width="964" height="724" alt="Image" src="https://github.com/user-attachments/assets/0af084d8-3645-4b4d-9637-a6675723a64c" />
- The title edit button (pen icon) has no description
- The aria-label for the title is weird with the Loading... `<a href="" aria-label="Loading..." data-l10n-aria-label="true">Test</a>` 

It is worth it to take a look at the other output from a lighthouse test as well!
